command/init.go

```diff
@@ -168,61 +167,34 @@ func (c *InitCommand) Run(args []string) int {
 		return 0
 	}
 
-	// Before we do anything else, we'll try loading configuration with both
-	// our "normal" and "early" configuration codepaths. If early succeeds
-	// while normal fails, that strongly suggests that the configuration is
-	// using syntax that worked in 0.11 but no longer in 0.12, which requires
-	// some special behavior here to get the directory initialized just enough
-	// to run "terraform 0.12upgrade".
-	//
-	// FIXME: Once we reach 0.13 and remove 0.12upgrade, we should rework this
-	// so that we first use the early config to do a general compatibility
+	// First use the early config to do a general compatibility
 	// check with dependencies, producing version-oriented error messages if
-	// dependencies aren't right, and only then use the real loader to deal
-	// with the backend configuration.
-	rootMod, confDiags := c.loadSingleModule(path)
+	// dependencies aren't right.
 	rootModEarly, earlyConfDiags := c.loadSingleModuleEarly(path)
-	configUpgradeProbablyNeeded := false
+	diags = diags.Append(earlyConfDiags)
+	if earlyConfDiags.HasErrors() {
+		// Errors from the early loader are generally not as high-quality since
+		// it has less context to work with.
+		c.Ui.Error(strings.TrimSpace(errInitConfigError))
+	}
```

Comment:
If I'm reading this correctly, if the early loader has errors and the real loader doesn't, we'll print an error but no diagnostics and won't exit. This seems like a basically-impossible situation to be in, but maybe we should try to handle it somehow anyway? We did before, I think.

Reply:
This prints the error, but waits till the next step to return so we can customize the response based on whether both config loaders fail, or just the latter. 

Having said that, I'm making a big assumption that the config loader will _always_ produce an error if the earlyconfig produces an error, and just because that's how it's _supposed_ to work doesn't mean that's how it _will_ work. 

I'm going to make a couple of small changes and re-request review, thanks for bringing this up!